website/docs/r/vapp_nat_rules.html.markdown

```diff
@@ -94,6 +92,7 @@ The following arguments are supported:
 * `vdc` - (Optional) The name of VDC to use, optional if defined at provider level.
 * `vapp_id` - (Required) The identifier of [vApp](/docs/providers/vcd/r/vapp.html).
 * `network_id` - (Required) The identifier of [vApp network](/docs/providers/vcd/r/vapp_network.html).
+* `enabled` - (Optional) Enable or disable NAT. Default is `true`.
```

Comment:
Opening up a discussion and sharing my findings on behavior.

1.) I was able to hit a race-condition and see the warning: 

```
Terraform will perform the following actions:

  # vcd_vapp_firewall_rules.vapp_fw will be updated in-place
  ~ resource "vcd_vapp_firewall_rules" "vapp_fw" {
        default_action     = "drop"
      ~ enabled            = false -> true
        id                 = "urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4"
        log_default_action = false
        network_id         = "urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4"
        vapp_id            = "urn:vcloud:vapp:9902c43c-78bf-41eb-a895-04c482b077cc"

      + rule {
          + destination_ip   = "any"
          + destination_port = "21"
          + enable_logging   = false
          + enabled          = true
          + name             = "demo-drop-ftp"
          + policy           = "drop"
          + protocol         = "tcp"
          + source_ip        = "10.10.0.0/24"
          + source_port      = "any"
        }
      + rule {
          + destination_ip   = "any"
          + destination_port = "any"
          + enable_logging   = false
          + enabled          = true
          + name             = "demo-allow-outbound"
          + policy           = "allow"
          + protocol         = "any"
          + source_ip        = "10.10.0.0/24"
          + source_port      = "any"
        }
    }

  # vcd_vapp_nat_rules.vapp_nat will be updated in-place
  ~ resource "vcd_vapp_nat_rules" "vapp_nat" {
        enable_ip_masquerade = false
      ~ enabled              = false -> true
        id                   = "urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4"
        nat_type             = "portForwarding"
        network_id           = "urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4"
        vapp_id              = "urn:vcloud:vapp:9902c43c-78bf-41eb-a895-04c482b077cc"

        rule {
            external_port   = 22
            forward_to_port = -1
            id              = "65537"
            protocol        = "TCP"
            vm_id           = "urn:vcloud:vm:95313037-ea9f-4725-9795-d694666e2a64"
            vm_nic_id       = 0
        }
        rule {
            external_port   = -1
            forward_to_port = 80
            id              = "65538"
            protocol        = "TCP"
            vm_id           = "urn:vcloud:vm:95313037-ea9f-4725-9795-d694666e2a64"
            vm_nic_id       = 0
        }
    }

Plan: 0 to add, 2 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

vcd_vapp_firewall_rules.vapp_fw: Modifying... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4]
vcd_vapp_nat_rules.vapp_nat: Modifying... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4]
vcd_vapp_nat_rules.vapp_nat: Still modifying... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4, 10s elapsed]
vcd_vapp_firewall_rules.vapp_fw: Still modifying... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4, 10s elapsed]
vcd_vapp_firewall_rules.vapp_fw: Still modifying... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4, 20s elapsed]
vcd_vapp_nat_rules.vapp_nat: Still modifying... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4, 20s elapsed]
WARNING: to enable NAT, Firewall has to be enabled. Please use vcd_vapp_firewall_rules to enable firewall.
vcd_vapp_firewall_rules.vapp_fw: Still modifying... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4, 30s elapsed]
vcd_vapp_nat_rules.vapp_nat: Still modifying... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4, 30s elapsed]
vcd_vapp_nat_rules.vapp_nat: Modifications complete after 35s [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4]
vcd_vapp_firewall_rules.vapp_fw: Still modifying... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4, 40s elapsed]
vcd_vapp_firewall_rules.vapp_fw: Still modifying... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4, 50s elapsed]
vcd_vapp_firewall_rules.vapp_fw: Still modifying... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4, 1m0s elapsed]
vcd_vapp_firewall_rules.vapp_fw: Modifications complete after 1m3s [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4]
```

But why, even though there was a warning, NAT still became enabled?

<img width="799" alt="Screenshot 2020-06-26 at 10 25 48" src="https://user-images.githubusercontent.com/11330689/85831755-717a3300-b797-11ea-8fb2-3109956d8f3b.png">

2.) Then I disabled the FW:

```
Terraform will perform the following actions:

  # vcd_vapp_firewall_rules.vapp_fw will be updated in-place
  ~ resource "vcd_vapp_firewall_rules" "vapp_fw" {
        default_action     = "drop"
      ~ enabled            = true -> false
        id                 = "urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4"
        log_default_action = false
        network_id         = "urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4"
        vapp_id            = "urn:vcloud:vapp:9902c43c-78bf-41eb-a895-04c482b077cc"

        rule {
            destination_ip        = "any"
            destination_port      = "21"
            destination_vm_nic_id = 0
            enable_logging        = false
            enabled               = true
            name                  = "demo-drop-ftp"
            policy                = "drop"
            protocol              = "tcp"
            source_ip             = "10.10.0.0/24"
            source_port           = "any"
            source_vm_nic_id      = 0
        }
        rule {
            destination_ip        = "any"
            destination_port      = "any"
            destination_vm_nic_id = 0
            enable_logging        = false
            enabled               = true
            name                  = "demo-allow-outbound"
            policy                = "allow"
            protocol              = "any"
            source_ip             = "10.10.0.0/24"
            source_port           = "any"
            source_vm_nic_id      = 0
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

vcd_vapp_firewall_rules.vapp_fw: Modifying... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4]
vcd_vapp_firewall_rules.vapp_fw: Still modifying... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4, 10s elapsed]
vcd_vapp_firewall_rules.vapp_fw: Still modifying... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4, 20s elapsed]
vcd_vapp_firewall_rules.vapp_fw: Still modifying... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4, 30s elapsed]
vcd_vapp_firewall_rules.vapp_fw: Modifications complete after 33s [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4]
```

To my surprise, NAT were still enabled!

<img width="867" alt="Screenshot 2020-06-26 at 10 29 56" src="https://user-images.githubusercontent.com/11330689/85832017-fd8c5a80-b797-11ea-99fb-c3b9618c7662.png">

However, UI was throwing a warning:

<img width="861" alt="Screenshot 2020-06-26 at 10 30 41" src="https://user-images.githubusercontent.com/11330689/85832081-18f76580-b798-11ea-9902-42a4a01b5081.png">

Do we want to throw our warning during a refresh too? Is it even possible? Because right now, FW is disabled, NAT is enabled, but there is no warning:

```
$ terraform apply
vcd_external_network.demo_ext_net: Refreshing state... [id=urn:vcloud:network:c8f62b1b-92c6-40f1-85d6-e3a387e06428]
vcd_org.demo_org: Refreshing state... [id=urn:vcloud:org:19361f4f-cd6f-416f-96a8-63a79824b56e]
vcd_org_vdc.demo_vdc: Refreshing state... [id=urn:vcloud:vdc:7ab03b09-1cc2-4566-bcd0-9c52549f4f6b]
vcd_edgegateway.demo_gw: Refreshing state... [id=urn:vcloud:gateway:4f238eca-953e-4863-ac7e-88f008bef4ec]
vcd_network_routed.demo_routed_net_web: Refreshing state... [id=urn:vcloud:network:19991f34-7922-4d9e-85a7-7dbdef93a165]
vcd_nsxv_firewall_rule.demo_nsxv_fw_rule: Refreshing state... [id=131078]
vcd_vapp.demo_vapp: Refreshing state... [id=urn:vcloud:vapp:9902c43c-78bf-41eb-a895-04c482b077cc]
vcd_vapp_org_network.demo_vapp_net_org: Refreshing state... [id=urn:vcloud:network:dcab8f24-81bf-4934-9d54-41707ea6b4c8]
vcd_vapp_network.demo_vapp_net_w_org: Refreshing state... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4]
vcd_vapp_firewall_rules.vapp_org_fw: Refreshing state... [id=urn:vcloud:network:dcab8f24-81bf-4934-9d54-41707ea6b4c8]
vcd_vapp_vm.demo_vm_wordpress: Refreshing state... [id=urn:vcloud:vm:95313037-ea9f-4725-9795-d694666e2a64]
vcd_vapp_static_routing.vapp_demo_static_routing1: Refreshing state... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4]
vcd_vapp_firewall_rules.vapp_fw: Refreshing state... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4]
vcd_nsxv_dnat.demo_dnat_wordpress_https1: Refreshing state... [id=196610]
vcd_vapp_nat_rules.vapp_nat: Refreshing state... [id=urn:vcloud:network:241144cd-a8b2-4efc-abad-b0fce8c4e4a4]
```

3.) So, do we need (or not) to add dependencies to examples like below?

```
  # Dependency on firewall to avoid a possible race condition and a warning
  # when enabling NAT rules, because FW rules need to be enabled first
  depends_on = [vcd_vapp_firewall_rules.vapp_fw]
```

It is not fully clear to me, because even without dependency, it seems, that the NAT rules still get enabled, even if they are enabled before FW. We just get a warning.
With dependency the apply is cleaner though.

Reply:
4.) Let's also add a sentence to the argument doc itself (bold only here - no need in docs):

> * `enabled` - (Optional) Enable or disable NAT. Default is `true`. **To enable the NAT service, firewall[link to vcd_vapp_firewall_rules] needs to be enabled as well.**

Reply:
added 4

Reply:
1. API allows to enable NAT even FW is disabled. It just in UI shows warning.

Reply:
2. Explained in 1

Reply:
3. depends isn't needed - just warning may showed if NAT first not FW.

Reply:
> added 4

It's a mess:

<img width="753" alt="Screenshot 2020-06-26 at 14 42 30" src="https://user-images.githubusercontent.com/11330689/85853505-49043000-b7bb-11ea-9df0-035e8aceb97e.png">

Reply:
improved

Reply:
> 3. depends isn't needed - just warning may showed if NAT first not FW.

OK, I tried ten times different ways with new code and I cannot hit the race condition anymore. It doesn't mean it's not there, but it's just not showing. I'm retracting this proposal. It would be required only if trying to enable NAT failed when firewall is not enabled, which is not the case. And the hassle of requiring `depends_on` is not balanced by the benefits (not showing what is only a warning in this rare race condition).